clouddrift/wavelet.py

```diff
@@ -0,0 +1,534 @@
+        wavelets must be last.
+
+    Returns
+    -------
+    wtx : np.ndarray
+        Time-domain wavelet transforms of input x. w shape will be ((series_orders), order, freq_axis, time_axis).
+
+    Examples
+    --------
+    To write.
+
+    Raises
+    ------
+    ValueError
+        If the time axis is outside of the valid range ([-1, N-1]).
+        If the shape of time axis is different for input and wavelet.
+        If ``boundary not in ["mirror", "zeros", "periodic"]``.
+
+    See Also
+    --------
+    :func:`morsewave`, `morsefreq`
+    """
+    # time_axis must be in valid range
+    if time_axis < -1 or time_axis > len(x.shape) - 1:
+        raise ValueError(
+            f"time_axis ({time_axis}) is outside of the valid range ([-1,"
+            f" {len(x.shape) - 1}])."
+        )
+    # Positions and time arrays must have the same shape.
+    if x.shape[time_axis] != wave.shape[-1]:
+        raise ValueError("x and wave time axes must have the same length.")
+
+    wave_ = np.moveaxis(wave, [freq_axis, order_axis], [-2, -3])
+
+    # initialization: output will be ((x_orders),f_order, freq_axis, time_axis)
+    # w = np.tile(
+    #     np.expand_dims(np.zeros_like(x), (-3, -2)),
+    #     (1, np.shape(wave)[-3], np.shape(wave)[-2], 1),
+    # )
+
+    # if x is of dimension 1 we need to expand
+    # otherwise make sure time axis is last
+    if np.ndim(x) < 2:
+        x_ = np.expand_dims(x, axis=0)
+    else:
+        x_ = np.moveaxis(x, time_axis, -1)
+
+    if ~np.all(np.isreal(x)):
+        if norm == "energy":
+            x_ /= np.sqrt(2)
+        elif norm == "bandpass":
+            x_ /= 2
+
+    # to do: add detrending option by default?
+
+    # apply boundary conditions
+    if boundary == "mirror":
+        x_ = np.concatenate((np.flip(x_, axis=-1), x_, np.flip(x_, axis=-1)), axis=-1)
+    elif boundary == "zeros":
+        x_ = np.concatenate((np.zeros_like(x_), x_, np.zeros_like(x_)), axis=-1)
+    elif boundary == "periodic":
+        x_ = np.concatenate((x_, x_, x_), axis=-1)
+    else:
+        raise ValueError("boundary must be one of 'mirror', 'align', or 'zeros'.")
+
+    # danger zone, use different letters from jlab: n is their time length
+    time_length = np.shape(x)[-1]
+    time_length_ = np.shape(x_)[-1]
+
+    # pad wavelet with zeros?
+    # wave_ = np.concatenate((np.zeros_like(wave), wave, np.zeros_like(wave)), axis=-1)
+    order_length, freq_length, _ = np.shape(wave)
+    wave_ = np.zeros((order_length, freq_length, time_length_), dtype="cfloat")
+    # display(np.shape(wave_))
+    index = slice(
+        int(np.floor(time_length_ - time_length) / 2),
+        int(time_length + np.floor(time_length_ - time_length) / 2),
+    )
+    wave_[:, :, index] = wave
+
+    # here we assume that wavelet and data have exactly the same number of data points;
+    # to do: include case wavelet is shorter or longer
+
+    # take fft along axis = -1
+    wavefft_ = np.fft.fft(wave_)
```

Comment:
From https://numpy.org/devdocs/reference/routines.fft.html#type-promotion, it looks like automatically, the output of `numpy.fft` is converted to `float64` and `complex128`. So I don't think you need to manually define `dtype="cfloat"` for `wave_`.

Reply:
ok! Then do you think we should bother to make sure we return the transform `wtx` with the same precision as the input `x`?

Reply:
In general, I would not modify the type of the output and leave it with the same precision as the input anywhere where it's possible.